### PR TITLE
fix broadcast on mp mode

### PR DESCRIPTION
**Pull request by Yancey0623**

### CODE REVIEW

**Review by typhoonzero — DISMISSED** [review]

**Comment by Yancey0623** [comment]
Need @panyx0718 to review the API changing.

File: paddle/fluid/framework/parallel_executor.cc [hunk 1]
```diff
@@ -254,9 +255,17 @@ ParallelExecutor::ParallelExecutor(
     PADDLE_THROW("Not compiled with CUDA");
 #endif
   }
-  if (member_->local_scopes_.size() != 1 && local_scopes.empty()) {
-    BCastParamsToDevices(bcast_vars);
+  // broadcast parameters from device 0 to others:
+  // 1. multiple trainer instances
+  // 2. one trainer with multiple devices
+  if (build_strategy.num_trainers_ > 1) {
+    BCastParamsToDevices(bcast_vars, build_strategy.trainer_id_);
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.cc]: does bcast support multiple trainers? Seems not good on cpu

File: paddle/fluid/framework/parallel_executor.cc [hunk 2]
```diff
@@ -254,9 +255,17 @@ ParallelExecutor::ParallelExecutor(
...
+  } else {
+    if (member_->local_scopes_.size() != 1 && local_scopes.empty()) {
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.cc]: else if?

File: paddle/fluid/framework/parallel_executor.h [hunk 3]
```diff
@@ -70,7 +72,8 @@ class ParallelExecutor {
            const std::string &fetched_var_name);
 
  private:
-  void BCastParamsToDevices(const std::unordered_set<std::string> &vars) const;
+  void BCastParamsToDevices(const std::vector<std::string> &vars,
+                            int trainer_id = 0) const;
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.h:78]: document the meaning of trainer_id?

File: paddle/fluid/framework/parallel_executor.cc (hunk 1 above)
Comment (Yancey0623) [reply paddle/fluid/framework/parallel_executor.cc]: Always `num_trainer=1, trainer_id=0` in CPU distributed training(pserver mode).

File: paddle/fluid/framework/parallel_executor.cc (hunk 2 above)
Comment (Yancey0623) [reply paddle/fluid/framework/parallel_executor.cc]: done.

File: paddle/fluid/framework/parallel_executor.h (hunk 3 above)
Comment (Yancey0623) [reply paddle/fluid/framework/parallel_executor.h:78]: done.

**Review by chengduoZH — APPROVED** [review]
LGTM

**Review by panyx0718 — APPROVED** [review]
LG compiler.py changes.

**Pull request merged**